### ISSUE TITLE
Disbale the browser navigaiton stack in flutter web app

### ISSUE THREAD

**Issue by kodejack** [flutter/flutter#108114]
Disable the browser navigation stack in the Flutter web app to allow a Flutter app to act more like a Single Page Application (SPA) and manage its navigation internally 
## Use case

We have an application for iOS, Android and Web. The app works perfectly on mobile platforms, but the browser back button causes mayhem for the web application. The application has many dynamically generated pages, which makes the browser history very difficult to manage. If the user uses the navigation in the flutter app, everything is good. If the user starts using the browser's back and forward navigation, The app will regularly fall over due to this being wrong. We have no requirement for deep linking and would use the app as if it was SPA and let Flutter manage the navigation separately from the browser. i.e. if you hit back on the browser, you leave the page.    

We have looked at various ways to manage or override this behaviour, including route guards, but have failed to find a reasonable solution. 
We also tried using w iFrame, which caused cross-site scripting issues with our authentication. 

We have written several web apps in Flutter, and all of them have experienced this issue in some way.

## Proposal

What we like is the ability to switch a flutter web app from a website with full navigation controlled by the browser and Flutter to a web app that handles its own navigation. Similar to it being contained in an iframe.  

If there is already a way to do this, I would love to know. Alas, I have not found a way to this as yet.

**Comment by maheshmnj** [comment]
Hi @kodejack, Thanks for filing the issue. This looks like a valid feature request. Labeling this issue for further insights from the team.

**Comment by maheshmnj** [comment]
cc: @mdebbar

**Comment by kodejack** [comment]
We looked at WillPop, but it doesn't seem to work correctly on web and didn't always suit our scenario.

**Comment by yjbanov** [comment]
Have you tried using a custom `UrlStrategy`?

https://api.flutter.dev/flutter/package-flutter_web_plugins_url_strategy/UrlStrategy-class.html

**Comment by kodejack** [comment]
I dont think this will fix the issue as we have a number of pages that are dynamic. All call "FormStack". These are generated on they fly depedning the form branching logic. I will have a closer look at this as it might some of the issue but not the dynamic pages.

**Comment by blaneyneil** [comment]
I'm running into this same problem with a soon-to-be-released web build of our app.  To make matters worse, our web version uses a separate Navigator key so we can target into different views.  Which makes a parent "back" inherently confusing.

It would be really helpful to have an option to disable the browser nav stack on the MaterialApp() level.

**Comment by VIP-Dev-Invisible** [comment]
I am also having the same issue . There should be an option in MaterialApp() where user can disable Nav stack. Is there any ETA to solve this issue ?

**Comment by mdebbar** [comment]
In order to disable all browser history, you could do something like this:
```dart
import 'package:flutter_web_plugins/url_strategy.dart';

void main() {
  // Before running your app:
  setUrlStrategy(null);

  runApp(...);
}
```
Of course, you'll need to add the `flutter_web_plugins` package to your `pubspec.yaml`:
```yaml
dependencies:
  flutter_web_plugins:
    sdk: flutter
```

@kodejack is this what you're looking for?

**Comment by kodejack** [comment]
@mdebbar this seems to be working as desired, The browser nav stack and hence the buttons are being ignored. This looks like we could be onto a winner.   I'll try a few other things and let you know if this is pretty solid.

**Comment by kodejack** [comment]
After trying a few things this seem to fix the web browser nav stack but it makes it incompatable with the iOS and Android apps assuming this is be cause of dependency on `flutter_web_plugins` Can these co-exist with mobile?

```yml
dependencies:
  flutter_web_plugins:
    sdk: flutter
```

**Comment by mdebbar** [comment]
@kodejack that's the right way to do it, yes. But we already do this for you in:
https://github.com/flutter/flutter/blob/fc444aedab1290b510ca85fc4a906e185cd3eaa6/packages/flutter_web_plugins/lib/url_strategy.dart#L7-L8

If you wanna take advantage of that, and keep your code simple, you can import `url_strategy.dart` directly instead of importing all of `flutter_web_plugins.dart`.

```dart
import 'package:flutter_web_plugins/url_strategy.dart';
```
This should work in mobile and web. On mobile, it's implemented as a no-op, similar to what you did in your code sample.

**Comment by janknips** [comment]
If anyone else comes across this, and needs to have the URL update, but no back button or history.

Just call setUrlStrategy() with the following:

```
class NoHistoryUrlStrategy extends PathUrlStrategy {
  @override
  void pushState(Object? state, String title, String url) =>
      replaceState(state, title, url);
}
```

**Comment by Tameflame** [comment]
@janknips how did you figure this out? This is exactly what I was looking for! Thanks!

**Comment by PriyaranjanEuphotic** [comment]
> @janknips how did you figure this out? This is exactly what I was looking for! Thanks!



> If anyone else comes across this, and needs to have the URL update, but no back button or history.
> 
> Just call setUrlStrategy() with the following:
> 
> ```
> class NoHistoryUrlStrategy extends PathUrlStrategy {
>   @override
>   void pushState(Object? state, String title, String url) =>
>       replaceState(state, title, url);
> }
> ```

its working, but one problem, when I am pressing back button its completely exiting from app, is there are way to display a confirmation dialog

**Comment by bobosette** [comment]
> In order to disable all browser history, you could do something like this:
> 
> ```dart
> import 'package:flutter_web_plugins/url_strategy.dart';
> 
> void main() {
>   // Before running your app:
>   setUrlStrategy(null);
> 
>   runApp(...);
> }
> ```
> 
> Of course, you'll need to add the `flutter_web_plugins` package to your `pubspec.yaml`:
> 
> ```yaml
> dependencies:
>   flutter_web_plugins:
>     sdk: flutter
> ```
> 
> @kodejack is this what you're looking for?

you saved my life!!!

**Comment by millarboy** [comment]
Yo, I am definitely under qualified to be commenting here, However I did have some issues with hiding the safari bar in my flutter mobile web app so in case I understood the issue i fixed it by  wrapping everything on the page with a non scrolling container.... it made my system nav go.... sorry if this is completely out of context

**Comment by adamcastleman** [comment]
Insert this script into your index.html file to disable browser gesture navigation on flutter web:

`<script>
  window.addEventListener('touchstart', handleTouchStart, false);        
  window.addEventListener('touchmove', handleTouchMove, {passive: false});

  let xDown = null;                                                        
  let yDown = null;

  function getTouches(evt) {
    return evt.touches || evt.originalEvent.touches;
  }                                                     

  function handleTouchStart(evt) {
      const firstTouch = getTouches(evt)[0];                                      
      xDown = firstTouch.clientX;                                      
      yDown = firstTouch.clientY;                                      
  };                                                

  function handleTouchMove(evt) {
      if (!xDown || !yDown) {
          return;
      }

      let xUp = evt.touches[0].clientX;                                    
      let yUp = evt.touches[0].clientY;

      let xDiff = xDown - xUp;
      let yDiff = yDown - yUp;

      if (Math.abs(xDiff) > Math.abs(yDiff)) {
          evt.preventDefault();
      } 

      // Reset values
      xDown = null;
      yDown = null;                                             
  };
</script>
`

**Comment by adamcastleman** [comment]
Insert this script into you index.html file in your web folder to disable mobile browser gesture swiping

`<script>
  window.addEventListener('touchstart', handleTouchStart, false);        
  window.addEventListener('touchmove', handleTouchMove, {passive: false});

  let xDown = null;                                                        
  let yDown = null;

  function getTouches(evt) {
    return evt.touches || evt.originalEvent.touches;
  }                                                     

  function handleTouchStart(evt) {
      const firstTouch = getTouches(evt)[0];                                      
      xDown = firstTouch.clientX;                                      
      yDown = firstTouch.clientY;                                      
  };                                                

  function handleTouchMove(evt) {
      if (!xDown || !yDown) {
          return;
      }

      let xUp = evt.touches[0].clientX;                                    
      let yUp = evt.touches[0].clientY;

      let xDiff = xDown - xUp;
      let yDiff = yDown - yUp;

      if (Math.abs(xDiff) > Math.abs(yDiff)) {
          evt.preventDefault();
      } 

      // Reset values
      xDown = null;
      yDown = null;                                             
  };
</script>
`
